### PR TITLE
discovery and print timeouts

### PR DESCRIPTION
**Pull request by israelko**
- Finish discovery when printer is found

In the current version, printer discovery may take a long time in android because the lib always waits to the timeout to come and only then sends back to the js the list of printers. 

So, I've added an optional param to the discover() function "findFirstAndroid" that indicates that the discovery should end and results return to js once the first printer was found.

- Print timeout

In the current version, the print command uses the default 5000 ms timeout which sometimes could not be enough (for example when sending multiple receipt images to the printer).

So, I've added an optional param to the printing.send() method to specify the print command timeout.

(sorry if I didn't follow a PR template, I couldn't find one for this repo).

### CODE REVIEW

File: android/src/main/java/com/reactnativeescposprinter/EscPosPrinterDiscoveryModule.java [hunk 1]
```diff
@@ -261,6 +273,14 @@ public void sendDiscoveredDataToJS() {
 
   @ReactMethod
   private void discover(final ReadableMap paramsMap, Promise promise) {
+    mFindFirst = false;
+
+    if (paramsMap != null) {
+      if (paramsMap.hasKey("findFirstAndroid")) {
```
Comment (tr3v3r) [inline android/src/main/java/com/reactnativeescposprinter/EscPosPrinterDiscoveryModule.java:279]: Can we extract please such params assignments ( findFirstAndroid, usbSerialNumber and etc. ) to a separate method called? smth like:

**defineSettingsFromParamsMap**

And then:
```
@ReactMethod
  private void discover(final ReadableMap paramsMap, Promise promise) {
    mFindFirst = false;

    defineSettingsFromParamsMap(paramsMap)


```

File: src/printing.ts [hunk 2]
```diff
@@ -497,8 +497,8 @@ class Printing {
     return this;
   }
 
-  send() {
-    return this._send(this._buffer);
+  send(timeout?: number) {
```
Comment (tr3v3r) [inline src/printing.ts:500]: I think we need to throw error here if timeout more/less the constraints from the Epson docs.
ie. (timeout < 5000 || timeout > 600000) for iOS 
throw new Error('Timeout should be an integer from 5000 to 600000*)')

File: ios/EscPosPrinter.m [hunk 3]
```diff
@@ -220,11 +220,12 @@ + (BOOL)requiresMainQueueSetup
 
 
 RCT_EXPORT_METHOD(printBuffer: (NSArray *)printBuffer
+                  timeout:(int)timeout
```
Comment (tr3v3r) [inline ios/EscPosPrinter.m:223]: I think it'll be better if we could pass an object here like `params` in discovery. And I suppose it should be optional.

File: android/src/main/java/com/reactnativeescposprinter/EscPosPrinterModule.java [hunk 4]
```diff
@@ -498,11 +497,11 @@ public void stopMonitorPrinter(Promise promise) {
   }
 
   @ReactMethod
-  public void printBuffer(ReadableArray printBuffer, Promise promise) {
+  public void printBuffer(ReadableArray printBuffer, int timeout, Promise promise) {
```
Comment (tr3v3r) [inline android/src/main/java/com/reactnativeescposprinter/EscPosPrinterModule.java:500]: I think it'll be better if we could pass an object here like params in discovery. And I suppose it should be optional.

File: android/src/main/java/com/reactnativeescposprinter/EscPosPrinterModule.java [hunk 5]
```diff
@@ -312,10 +312,9 @@ private void disconnectPrinter() {
       mPrinter.clearCommandBuffer();
     }
 
-    private void printData() throws Epos2Exception {
+    private void printData(int timeout) throws Epos2Exception {
       this.connectPrinter();
-      mPrinter.sendData(Printer.PARAM_DEFAULT);
-
+      mPrinter.sendData(timeout > 5000 ? timeout : Printer.PARAM_DEFAULT);
```
Comment (tr3v3r) [inline android/src/main/java/com/reactnativeescposprinter/EscPosPrinterModule.java:317]: With all further comments I've passed, I suppose it'll better to use it like 
```
mPrinter.sendData(timeout || Printer.PARAM_DEFAULT);
```

File: docs/PRINTING.md [hunk 6]
```diff
@@ -202,6 +202,8 @@ printing.initialize().addPulse('EPOS2_DRAWER_2PIN');
 
 Is required at the end of a printer chain to send the commands to the printer
 
+- timeout? : the print operation timeout in ms (5000 - 300000). Optional number parameter, defaults to 5000.
```
Comment (tr3v3r) [inline docs/PRINTING.md:205]: for iOS I suppose default is 10000

File: android/src/main/java/com/reactnativeescposprinter/EscPosPrinterDiscoveryModule.java (hunk 1 above)
Comment (israelko) [reply android/src/main/java/com/reactnativeescposprinter/EscPosPrinterDiscoveryModule.java:279]: will do.

File: src/printing.ts (hunk 2 above)
Comment (israelko) [reply src/printing.ts:500]: will do

File: ios/EscPosPrinter.m (hunk 3 above)
Comment (israelko) [reply ios/EscPosPrinter.m:223]: will do

File: android/src/main/java/com/reactnativeescposprinter/EscPosPrinterModule.java (hunk 4 above)
Comment (israelko) [reply android/src/main/java/com/reactnativeescposprinter/EscPosPrinterModule.java:500]: will do

File: android/src/main/java/com/reactnativeescposprinter/EscPosPrinterModule.java (hunk 5 above)
Comment (israelko) [reply android/src/main/java/com/reactnativeescposprinter/EscPosPrinterModule.java:317]: will be changed as I will switch to use a params object

File: docs/PRINTING.md (hunk 6 above)
Comment (israelko) [reply docs/PRINTING.md:205]: right :)

File: android/src/main/java/com/reactnativeescposprinter/EscPosPrinterDiscoveryModule.java [hunk 7]
```diff
@@ -295,4 +284,21 @@ public void onDone(String result) {
       }
     }, paramsMap);
   }
+
+  private void defineSettingsFromParamsMap(ReadableMap paramsMap) {
+    mExtractUsbSerialNumber = false;
+    mFindFirst = false;
+
+    if (paramsMap != null) {
+      if (paramsMap.hasKey("findFirstAndroid")) {
```
Comment (tr3v3r) [inline android/src/main/java/com/reactnativeescposprinter/EscPosPrinterDiscoveryModule.java:293]: I think scanningTimeoutAndroid also should be here

**Pull request merged**